log4j-api/src/main/java/org/apache/logging/log4j/ScopedContext.java

```diff
@@ -0,0 +1,225 @@
+     * @param key the key.
+     * @param value the value associated with the key.
+     * @param executorService the ExecutorService to dispatch the work.
+     * @param task the Callable to call.
+     */
+    public static <R> Future<R> callWhere(
+            final String key, final Object value, final ExecutorService executorService, final Callable<R> task) {
+        return provider.newScopedContext(key, value).call(executorService, task);
+    }
+
+    /**
+     * Creates a ScopedContext with a Map of keys and values and calls a method.
+     * @param map the Map.
+     * @param task the Runnable to call.
+     */
+    public static <R> R callWhere(final Map<String, ?> map, final Callable<R> task) throws Exception {
+        return provider.newScopedContext(map).call(task);
+    }
+
+    /**
+     * Return the object with the specified key from the current context.
+     * @param key the key.
+     * @return the value of the key or null.
+     * @param <T> The type of object expected.
+     * @throws ClassCastException if the specified type does not match the object stored.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> T get(String key) {
+        return (T) provider.getValue(key);
+    }
+
+    /**
+     * Return String value of the key from the current ScopedContext, if there is one and the key exists.
+     * @param key The key.
+     * @return The value of the key in the current ScopedContext.
+     */
+    public static String getString(String key) {
+        return provider.getString(key);
+    }
+
+    /**
+     * A holder of scoped context data.
+     */
+    public interface Instance {
+
+        /**
+         * Adds a key/value pair to the ScopedContext being constructed.
+         *
+         * @param key   the key to add.
+         * @param value the value associated with the key.
+         * @return the ScopedContext being constructed.
+         */
+        Instance where(String key, Object value);
+
+        /**
+         * Adds a key/value pair to the ScopedContext being constructed.
+         *
+         * @param key      the key to add.
+         * @param supplier the function to generate the value.
+         * @return the ScopedContext being constructed.
+         */
+        Instance where(String key, Supplier<Object> supplier);
+
+        /**
+         * Executes a code block that includes all the key/value pairs added to the ScopedContext.
+         *
+         * @param task the code block to execute.
+         */
+        void run(Runnable task);
+
+        /**
+         * Executes a code block that includes all the key/value pairs added to the ScopedContext on a different Thread.
+         *
+         * @param task the code block to execute.
+         * @return a Future representing pending completion of the task
+         */
+        Future<Void> run(ExecutorService executorService, Runnable task);
```

Comment:
I am not sure if we are not missing some use cases by not providing a:

```java
Runnable wrap(Runnable task);
```

method. For example with the available API we must write:

```java
ScopedContext.Instance scopedContext = ...;
asyncContext.startAsync(() -> scopedContext.run(() -> doSomething());
```

(where `asyncContext` is an [`AsyncContext`](https://jakarta.ee/specifications/servlet/6.0/apidocs/jakarta.servlet/jakarta/servlet/asynccontext#start(java.lang.Runnable))). With a `wrap` method we can do it with one less lambda:

```java
ScopedContext.Instance scopedContext = ...;
asyncContext.startAsync(scopedContext.wrap(() -> doSomething());
```

Note that the `run` and `call` methods could still be provided as shortcuts, but can have default implementations:

```java
default Future<Void> run(ExecutorService executorService, Runnable task) {
    executorService.submit(wrap(task), null);
}
```

Reply:
In essense you are saying that we need to allow the user to continue to create their own Threads and be able to pass the context to it instead of us wrapping the thread creation. Yeah, that makes sense.